Sprint 12 History

Nadezhda-Gurova opened the pull request.

Nadezhda-Gurova committed:
Revert "Revert "Спринт 12""

This reverts commit 88a0eb15d6a80b5c9de004d10f14700a1a23127a.

Nadezhda-Gurova committed:
Revert "Revert "Added work with history""

This reverts commit 08da35780e0a60b21e9ab9bde88d11c0f97eba98.

Nadezhda-Gurova committed:
Revert "Revert "clear history button added""

This reverts commit 76a1942504360ad7df4550b4807d46beddb4760a.

kgornostaeva commented on app/src/main/java/com/example/playlistmaker/Observable.kt:
```diff
@@ -0,0 +1,10 @@
+import com.example.playlistmaker.data.Track
+import java.util.Observer
+
+interface Observable {
```
🍏 Не используется, можно удалить

kgornostaeva commented on app/src/main/java/com/example/playlistmaker/activity/SettingsActivity.kt:
```diff
@@ -53,7 +75,7 @@ class SettingsActivity : AppCompatActivity() {
     }
 
-    private fun writeToSupport() {
+    private fun initWriteToSupport() {
```
🍏 Вызов startActivity может завершиться ошибкой. В основном, это происходит, когда мы пытаемся запустить стороннее приложение. Если система не сможет найти на устройстве пользователя приложение, подходящее под переданный нами intent, то выбросит exception. Обычно стандартные интенты обрабатываются стандартными приложениями, которые в большинстве случаев есть на устройствах пользователей, Но полагаться на это не стоит. Лучше обернуть вызов startActivity для открытия стороннего приложения в блок try-catch.  К примеру, сейчас твое приложение падает на некоторых эмуляторах при попытке обратиться в поддержку с экрана настроек.

kgornostaeva commented on app/src/main/java/com/example/playlistmaker/SearchHistory.kt:
```diff
@@ -0,0 +1,65 @@
+import com.example.playlistmaker.data.Track
+import com.google.gson.Gson
+
+class SearchHistory (private val sharedPreferences: SharedPreferences) {
```
🍏 Немного поехало форматирование. Для форматирования кода и удаления неиспользуемых импортов применяй следующие сочетания клавиш:
Windows: Control+Alt+L
MacOS: Command+Option+L
Больше горячих клавиш можно найти здесь: https://developer.android.com/studio/intro/keyboard-shortcuts

kgornostaeva commented:
⚠️ Если у пользователя пустая история, то он все равно видит кнопку «Очистить историю» и заголовок «Вы искали» 

<img width="240" alt="Снимок экрана 2023-07-28 в 11 06 02" src="https://github.com/Nadezhda-Gurova/Playlist_Maker/assets/48408134/43a350c8-3f73-4754-bf48-8624ccd069ef">

Nadezhda-Gurova committed:
- исправила порядок добавления треков
- кнопка "Очистить историю" остается на экране
- если у пользователя пустая история, «Очистить историю» и заголовок «Вы искали» - исчезают
- закруглила обложки треков

Nadezhda-Gurova committed:
обернула вызовы startActivity для открытия сторонних приложений в блоки try-catch

kgornostaeva commented on app/src/main/java/com/example/playlistmaker/activity/SettingsActivity.kt:
```diff
@@ -20,11 +21,22 @@ class SettingsActivity : AppCompatActivity() {
+        try {
+            initShare()
+        } catch (_: ActivityNotFoundException) {
+        }
```
🍏 При желании можно еще показывать пользователю сообщение о том, что действие нельзя выполнить. Например, с помощью тоста

kgornostaeva commented on app/src/main/java/com/example/playlistmaker/SearchHistory.kt:
```diff
@@ -1,65 +1,57 @@
-                .putString(VIEWED_TRACK, json)
-                .apply()
-    }
+    class Impl(
```
🍏 Имплементацию интерфейса рекомендую вынести в отдельный файл, чтобы не связывать интерфейс и реализацию так крепко. Реализаций интерфейса может быть несколько. 
В будущем они вообще будут разнесены в разные слои, когда вы познакомитесь с чистой архитектурой.

kgornostaeva commented:
<img width="242" alt="Снимок экрана 2023-08-03 в 10 59 04" src="https://github.com/Nadezhda-Gurova/Playlist_Maker/assets/48408134/39abaef7-ad38-4c53-9e37-28a9e18143c5">

kgornostaeva commented:
<img width="281" alt="Снимок экрана 2023-08-03 в 11 06 26" src="https://github.com/Nadezhda-Gurova/Playlist_Maker/assets/48408134/cccbed03-39bf-4c5e-ae9a-ff00ea3ec15b">

kgornostaeva commented:
Пользователь начал вводить запрос, но все еще видит историю поиска без кнопки "Очистить историю":
<img width="281" alt="Снимок экрана 2023-08-03 в 11 07 09" src="https://github.com/Nadezhda-Gurova/Playlist_Maker/assets/48408134/ed2c1541-90f7-4bf3-b833-f0445a94ce0d">

Nadezhda-Gurova committed:
исправила замечания:
- При нажатии на кнопку «Очистить историю», история просмотренных треков удаляется и исчезают список треков, кнопка «Очистить историю», заголовок «Вы искали»
- Если у пользователя плохой интернет, высвечивается только сообщение об ошибке
-  Если у пользователя пустая история, информация об истории не высвечивается

Nadezhda-Gurova committed:
- удаленный файл
